buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -1066,6 +1289,74 @@ public int capacity() {
         }
     }
 
+    private static final class SizeClassedChunk extends Chunk {
+        private static final int FREE_LIST_EMPTY = -1;
+        private final int segmentSize;
+        private final MpscIntQueue freeList;
+
+        SizeClassedChunk(AbstractByteBuf delegate, Magazine magazine, boolean pooled, int segmentSize,
+                         ChunkReleasePredicate shouldReleaseChunk) {
+            super(delegate, magazine, pooled, shouldReleaseChunk);
+            int capacity = delegate.capacity();
+            this.segmentSize = segmentSize;
+            int segmentCount = capacity / segmentSize;
+            assert segmentCount > 0: "Chunk must have a positive number of segments";
+            freeList = MpscIntQueue.create(segmentCount, FREE_LIST_EMPTY);
+            freeList.fill(segmentCount, new IntSupplier() {
+                int counter;
+                @Override
+                public int getAsInt() {
+                    return counter++;
+                }
+            });
+        }
+
+        @Override
+        public void readInitInto(AdaptiveByteBuf buf, int size, int startingCapacity, int maxCapacity) {
+            int segmentId = freeList.poll();
+            if (segmentId == FREE_LIST_EMPTY) {
+                throw new IllegalStateException("Free list is empty");
+            }
+
+            int startIndex = segmentId * segmentSize;
```

Comment:
if segmentSize is power of 2 this can be made a shift

Reply:
They're not always pow2, so then it'd be conditional and I don't think that's worth it. Integer multiplication is like 3-5 cycles overhead, tops, if I'm not mistaken.